File: src/za/co/adhd_developers/questions/Question37.java
package za.co.adhd_developers.questions;

import za.co.adhd_developers.tools.Question;
import za.co.adhd_developers.tools.Utils;

import java.util.Hashtable;

public class Question37 implements Question
{
    private Hashtable<String, Boolean> isPrime = new Hashtable<>();

    private int truncatablePrimeCount = 0;

    private long answer = 0;
    
    @Override
    public void doWork()
    {
        int currNumber = 0;
        while (this.truncatablePrimeCount < 11)
        {
            currNumber++;
            String currNumberStr = String.valueOf(currNumber);

            if (!isLocalPrime(currNumberStr))
            {
                continue;
            }

            Boolean leftToRight = true;
            for (int i = 0; i < currNumberStr.length(); i++)
            {
                String tmp = currNumberStr.substring(0,currNumberStr.length()-i);

                leftToRight = isLocalPrime(tmp);

                if (!leftToRight)
                {
                    break;
                }
            }

            if (!leftToRight)
            {
                continue;
            }

            Boolean rightToLeft = true;
            for (int i = 0; i < currNumberStr.length(); i++)
            {
                String tmp = currNumberStr.substring(i);

                rightToLeft = isLocalPrime(tmp);

                if (!rightToLeft)
                {
                    break;
                }
            }

            if (!rightToLeft)
            {
                continue;
            }

            if (currNumber == 2 || currNumber == 3 || currNumber == 5 || currNumber == 7)
            {
                continue;
            }

            this.truncatablePrimeCount++;
            this.answer += Long.valueOf(currNumber);
        }
//        for (String item : this.truncatablePrime)
//        {
//            System.out.println(item);
//        }


    }

    private boolean isLocalPrime(String number)
    {
        if (this.isPrime.containsKey(number))
        {
            return this.isPrime.get(number);
        }

        if (Utils.isPrime(Long.valueOf(number)))
        {
            this.isPrime.put(number, true);
            return true;
        }

        this.isPrime.put(number, false);
        return false;
    }

    @Override
    public void printAnswer()
    {
        System.out.println("=================================");
        System.out.println("Question 37");
        System.out.println("The sum of prime numbers that are truncatable from left to right and right to left, is: " + this.answer);
        System.out.println("=================================");
    }
}

File: src/za/co/adhd_developers/kotlin/questions/Question37.kt
package za.co.adhd_developers.kotlin.questions

import za.co.adhd_developers.kotlin.Question
import za.co.adhd_developers.kotlin.tools.isPrime

class Question37 : Question {

    var answer: Long = 0

    override fun doWork() {

        var count = 0
        var currNumber: Long = 10L

        while (count < 11) {
            if (isPrime(currNumber)) {
                var checkingNumber = currNumber.toString()
                checkingNumber = checkingNumber.removeRange(0,1)

                var ltr = true

                while (checkingNumber.isNotEmpty()) {
                    if (!isPrime(checkingNumber.toLong())) {
                        ltr = false
                        checkingNumber = ""
                    } else {
                        checkingNumber = checkingNumber.removeRange(0, 1)
                    }
                }

                var rtl = true

                checkingNumber = currNumber.toString()
                checkingNumber = checkingNumber.substring(0,checkingNumber.length-1)

                while (checkingNumber.isNotEmpty()) {
                    if (!isPrime(checkingNumber.toLong())) {
                        rtl = false
                        checkingNumber = ""
                    } else {
                        checkingNumber = checkingNumber.substring(0, checkingNumber.length - 1)
                    }
                }

                if (ltr && rtl) {
                    count++
                    this.answer = this.answer.plus(currNumber)
                }
            }
            currNumber++
        }
    }

    override fun printAnswer() {
        println("=================================")
        println("Question 37")
        println("The sum of prime numbers that are truncatable from left to right and right to left, is: " + this.answer)
        println("=================================")
    }
}